File: LinkedLists/SLL.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class SinglyLinkedList {
  constructor() {
    this.head = this.tail = null;
    this.length = 0;
  }
  append(value) {
    const newNode = new Node(value);
    if (!this.head) {
      this.head = this.tail = newNode;
    } else {
      this.tail.next = newNode;
      this.tail = newNode;
    }
    this.length++;
    return this;
  }
  prepend(value) {
    const newNode = new Node(value);
    if (!this.head) {
      this.head = this.tail = newNode;
    } else {
      newNode.next = this.head;
      this.head = newNode;
    }
    this.length++;
    return this;
  }
  insert(index, value) {
    if (index >= this.length) {
      return this.append(value);
    } else if (index === 0) {
      return this.prepend(value);
    }
    const newNode = new Node(value);
    const nodeToChange = this.traverseToIndex(index - 1);
    newNode.next = nodeToChange.next;
    nodeToChange.next = newNode;
    this.length++;
    return this;
  }

  traverseToIndex(index) {
    let currentNode = this.head;
    let counter = 0;
    while (counter !== index) {
      currentNode = currentNode.next;
      counter++;
    }
    return currentNode;
  }

  remove(index) {
    if (index >= this.length || index < 0) {
      return this;
    }
    if (this.length === 1) {
      this.head = this.tail = null;
    } else if (index === 0) {
      this.head = this.head.next;
    } else {
      const preceedingNode = this.traverseToIndex(index - 1);
      const nodeToRemove = preceedingNode.next;
      if (nodeToRemove === this.tail) {
        this.tail = preceedingNode;
        preceedingNode.next = null;
      } else {
        preceedingNode.next = nodeToRemove.next;
      }
    }
    this.length--;
    return this;
  }

  printList() {
    const linkedArray = [];
    let currentNode = this.head;
    while (currentNode !== null) {
      linkedArray.push(currentNode.value);
      currentNode = currentNode.next;
    }
    return linkedArray;
  }
}

let myLinkedList = new SinglyLinkedList();
myLinkedList.append(4).append("second");
myLinkedList.append({ a: true });
myLinkedList.append(["man"]).prepend("First now");
//['first now',4,'second',{a:true},'man']
myLinkedList.insert(5, "index 1").insert(0, "index 0");
//['index 0' ,'first now',4,'second',{a:true},'man','index 1']
myLinkedList.insert(3, "third index");
//['index 0' ,'first now',4,'third index','second',{a:true},'man','index 1']

// console.log(myLinkedList.printList());
// console.log(myLinkedList);
// console.log(myLinkedList.printList());
// myLinkedList.remove(3);
// console.log(myLinkedList.printList());
// console.log(myLinkedList);
//['first now',4,'third index','second',{a:true},'man','index 1']

File: DSA-Python/SLL.py
class Node:
    def __init__(self,data) -> None:
        self.data = data
        self.next = None
        
class SinglyLinkedlist:
    def __init__(self) -> None:
        self.head,self.tail = None,None
        self.size = 0
    
    def append(self,data):
        new_node = Node(data)
        
        if self.head == None:
            self.head,self.tail = new_node,new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
        
        self.size += 1
        return self
    
    def prepend(self,data):
        new_node = Node(data)
        
        if self.head == None:
            self.head,self.tail = new_node,new_node
        else:
            new_node.next = self.head
            self.head = new_node
        
        self.size += 1
        return self

    def insert(self,data,index):
        if index >= self.size:
            return self.append(data)
        elif index == 0:
            return self.prepend(data)
        
        new_node = Node(data)
        node_to_change = self.traverseToIndex(index-1)
        
        new_node.next = node_to_change.next
        node_to_change.next = new_node
        self.size += 1
        
        return self
        
    def traverseToIndex(self,index):
        current_node = self.head
        counter = 0
        
        while counter != index:
            current_node = current_node.next
            counter += 1
            
        return current_node
    
    def remove(self,index):
        if index >= self.size or index < 0 :
            return self
        
        if self.size == 1:
            self.head,self.tail = None,None
        elif index == 0:
            self.head = self.head.next
        else:
            preceeding_node = self.traverseToIndex(index-1)
            node_to_remove = preceeding_node.next
            
            if node_to_remove == self.tail:
                preceeding_node.next = None
                self.tail = preceeding_node
            else:
                preceeding_node.next = node_to_remove.next
        
        self.size -= 1
        return self
    
    def printList(self) :
        linked_list_array = []
        current_node = self.head
        while current_node != None:
            linked_list_array.append(current_node.data)
            current_node = current_node.next
        return linked_list_array
    
    def reverse(self):
        if self.head.next == None:
            return self
        
        first = self.head
        second = first.next
        
        while(second) :
            temp = second.next
            second.next = first
            first = second
            second = temp
        
        temp = self.tail
        self.tail = self.head
        self.head = temp
        self.tail.next = None
        
        return self
           
            
        


linked_list = SinglyLinkedlist()
linked_list.append("first").prepend("new first").insert("middle",1).remove(1).append("another")
print(linked_list.printList())
print(linked_list.reverse().printList())